### ISSUE TITLE
User dataset loading all the time and users/batches tab is not loading

### ISSUE THREAD

**Issue by kristijanzajakovskiodido** [SalesforceLabs/OrgCheck#413]
**Describe the bug**
We are facing issues with loading users and batches tab. We have 4.3M active users in our org (Production). After some time of loading page is failing and we need to refresh (nothing showed). In another org (Sandbox) with 3.5M users again is not working (only loading all the time).

**To Reproduce**
Dataset [users]: starting...  - it never get's finished.
![Screenshot_262](https://github.com/SalesforceLabs/OrgCheck/assets/165299358/2a916613-1a4d-4a62-95a9-df9c8f529bfd)

**Comment by VinceFINET** [comment]
Hello

We apologize for this issue you are having.
We have other users with the same issue specially when there are A LOT of active users in an org.

Can you run these queries (from 1 to 6) for me to better understand the situation please and put the results (only counts) in reply?

For query 7 you can DM me them on the Org Check slack (to subscribe follow https://sfdc.co/OrgCheck-Community)

**1. Count of all active users with a valid profile just like in the app**

_1.1 Query 1 (like in the app without the inner queries)_
```
SELECT COUNT(Id)
FROM User
WHERE Profile.Id != NULL 
AND IsActive = TRUE
```

_1.2 Query 2 (like the query 1 but where statement are in diff order just to check if the perf coud be better)_
```
SELECT COUNT(Id)
FROM User
WHERE IsActive = TRUE
AND Profile.Id != NULL 
```

_1.3 Query 3 (like the query 2 but we directly used profileid instead of the pointer)_
```
SELECT COUNT(Id)
FROM User
WHERE IsActive = TRUE
AND ProfileId != NULL 
```

Question: is query 2 or 3 faster than query 1 in your org?


**2. Count of all active users that are internal (we could think that skipping the external users make sense somehow)**
_2.1 Query 4 (external users have ContactId != null)_
```
SELECT COUNT(Id)
FROM User
WHERE IsActive = TRUE
AND ContactId != NULL 
AND ProfileId != NULL 
```

Question: is the query 4 even faster than query 1, 2 or 3 in your org?


**3. Finally, how many permission set assignment do you have per user i wonder?**
_3.1 Query 5 Count all Perm Set assignments for all users_
```
SELECT COUNT(Id)
FROM PermissionSetAssignment
WHERE Assignee.IsActive = TRUE
AND Assignee.ProfileId != NULL  
```

_3.2 Query 6 Count all Perm Set assignments for internal users_
```
SELECT COUNT(Id)
FROM PermissionSetAssignment
WHERE Assignee.IsActive = TRUE
AND Assignee.ProfileId != NULL  
AND Assignee.ContactId != NULL 
```

_3.3 Query 7 Count Perm Set assignments for each active all user_
As this query will bring a lot of data you can send it to me via 
slack (salesforce id and count does not contains any personal
information! don't worry)
```
SELECT AssigneeId, Assignee.ContactId, COUNT(Id)
FROM PermissionSetAssignment
WHERE Assignee.IsActive = TRUE
AND Assignee.ProfileId != NULL   
GROUP BY AssigneeId, Assignee.ContactId
```

**Comment by kristijanzajakovskiodido** [comment]
1. Count of all active users with a valid profile just like in the app

1.1 Query 1 (like in the app without the inner queries)

SELECT COUNT(Id)
FROM User
WHERE Profile.Id != NULL 
AND IsActive = TRUE

** COUNT = 4339280 **

1.2 Query 2 (like the query 1 but where statement are in diff order just to check if the perf coud be better)
(faster than query 1)
SELECT COUNT(Id)
FROM User
WHERE IsActive = TRUE
AND Profile.Id != NULL

** COUNT = 4339286 **
 
1.3 Query 3 (like the query 2 but we directly used profileid instead of the pointer)

SELECT COUNT(Id)
FROM User
WHERE IsActive = TRUE
AND ProfileId != NULL 
Question: is query 2 or 3 faster than query 1 in your org?

** COUNT = 4339305 **
** On first try the third query was slower than 1 and 2, but after one run it was much faster than query 1 and 2 ** 

2. Count of all active users that are internal (we could think that skipping the external users make sense somehow)
2.1 Query 4 (external users have ContactId != null)

SELECT COUNT(Id)
FROM User
WHERE IsActive = TRUE
AND ContactId != NULL 
AND ProfileId != NULL 
Question: is the query 4 even faster than query 1, 2 or 3 in your org?

** COUNT = 3827276 **
** 4th query is slower than query 3 **

3. Finally, how many permission set assignment do you have per user i wonder?
3.1 Query 5 Count all Perm Set assignments for all users

SELECT COUNT(Id)
FROM PermissionSetAssignment
WHERE Assignee.IsActive = TRUE
AND Assignee.ProfileId != NULL  

** COUNT = 4447919 **

3.2 Query 6 Count all Perm Set assignments for internal users

SELECT COUNT(Id)
FROM PermissionSetAssignment
WHERE Assignee.IsActive = TRUE
AND Assignee.ProfileId != NULL  
AND Assignee.ContactId != NULL 

** COUNT = 3915172 **

3.3 Query 7 Count Perm Set assignments for each active all user
As this query will bring a lot of data you can send it to me via
slack (salesforce id and count does not contains any personal
information! don't worry)

SELECT AssigneeId, Assignee.ContactId, COUNT(Id)
FROM PermissionSetAssignment
WHERE Assignee.IsActive = TRUE
AND Assignee.ProfileId != NULL   
GROUP BY AssigneeId, Assignee.ContactId

**Regarding this last query, this is the error that we are receiving while trying to execute "EXCEEDED_ID_LIMIT: Aggregate query does not support queryMore(), use LIMIT to restrict the results to a single batch" **

**Comment by VinceFINET** [comment]
so:
- internal and external users
   - 4.339k active internal and external users in this org (=query 3)
   - 4.448k permission set/profile assignment to active internal and external users (=query 5)
- only internal users
   - 3.827k active internal users in this org (=query 4)
   - 3.915k permission set/profile assignment to active internal users (=query 6)

most of your users don't have permission set assigned, just their profile.
- all active users: 97,54 % have only a profile and zero permset
- only internal users: 97,75 % have only a profile and zero permset

**Comment by VinceFINET** [comment]
for sure the code needs to be improved.
in your case, i am retrieving probably the same profile all over again for each user, where I could do this separatly in a much faster way.

**Comment by kristijanzajakovskiodido** [comment]
Hi Vince,

When we can expect fix for this issue on our orgs?

Best regards,
Kristijan Zajakovski

**Comment by kristijanzajakovskiodido** [comment]
Hi @VinceFINET, 

Any updates on this?

Best regards,
Kristijan

**Comment by VinceFINET** [comment]
Hello

This issue is under our radar but requires a refactoring on how we retrieve
perm set assignments for users
Specifically when the org has more than 1k users.
You will need to wait for Beryllium to be available on the appexchange.
Estimated date : July 2024

Sorry about this.

As a workaround I can provide you a wait to perform the same via sf cli.

**Comment by kristijanzajakovskiodido** [comment]
Hi Vincent,

As we chatted, we will wait for the upgrade in July. For now the task is postponed till July.

Best regads,
Kristijan Zajakovski

**Comment by VinceFINET** [comment]
limit the active users that are INTERNAL (so ContactID <> NULL) only 
switch the systempermission from User to Profile/PermSet so that it is retrieved only once per profile/permset (and not for each assignment !!!)
now profile and permsets have columns that shows if the open four critic system permissions.
as well for the users.

**Comment by VinceFINET** [comment]
optimizing as well the processing on the recipe level

**Comment by kristijanzajakovskiodido** [comment]
Hi Vincent, 

When you expect this to be applied on our orgs or ready to be installed?

Best regards,
Kristijan Zajakovski

**Comment by kristijanzajakovskiodido** [comment]
Hi Vince,

Do we have any updates on this?

Best regards,
Kristijan Zajakovski

**Comment by VinceFINET** [comment]
Hello

Can you make sure you have the latest version of Org Check from the
AppExchange?

It should be Beryllium.

**Comment by VinceFINET** [comment]
Can you confirm the version that you have?
If still on the Lithium version could you install from the AppExchange the new package?
thanks.

<img width="1328" alt="image" src="https://github.com/user-attachments/assets/713b4158-6a63-44a4-b9c3-64f1ee4307ee">

**Comment by kristijanzajakovskiodido** [comment]
Hi Vince,

I didn't have time to check but thank you for the update.

I will check end of this week or start of next week.

Best regards,
Kristijan Zajakovski
